server : add handle_slot_type lambda

danbev opened the pull request:
This commit adds a lambda function by extracting the common code for saving and restoring slots.

The motivation for this change is to reduce code duplication between the current two lambda expressions/functions.

danbev committed:
server : add handle_slot_type lambda

This commit adds a lambda function by extracting the common code for
saving and restoring slots.

The motivation for this change is to reduce code duplication between the
current two lambda expressions/functions.

ngxson commented on examples/server/server.cpp:
```diff
@@ -3665,7 +3665,8 @@ int main(int argc, char ** argv) {
     };
 
-    const auto handle_slots_save = [&ctx_server, &res_error, &res_ok, &params](const httplib::Request & req, httplib::Response & res, int id_slot) {
+    const auto handle_slot_type = [&ctx_server, &res_error, &res_ok, &params](const httplib::Request & req,
```
```suggestion
    const auto handle_slot_impl = [&ctx_server, &res_error, &res_ok, &params](const httplib::Request & req,
```

Rename to bring it more aligned with `handle_completions_impl` and `handle_embeddings_impl`

danbev replied:
I've renamed this now in https://github.com/ggerganov/llama.cpp/pull/11535/commits/7c34af40fb77807d6f06d40e4e7aa4e0437f396b.

ngxson commented on examples/server/server.cpp:
```diff
@@ -3691,37 +3692,18 @@ int main(int argc, char ** argv) {
         }
 
+        if (type == SERVER_TASK_TYPE_SLOT_SAVE) {
+            GGML_ASSERT(dynamic_cast<server_task_result_slot_save_load*>(result.get()) != nullptr);
```
missing check for `SERVER_TASK_TYPE_SLOT_RESTORE`

ngxson replied:
you I'm not sure if we can use `template` here to avoid having a long list of `if..else`

danbev replied:
I'll take a look at this :+1:

danbev replied:
I've added a suggestion for this in https://github.com/ggerganov/llama.cpp/pull/11535/commits/6cc2956f3fa6f387c1b4c2b99dc42720623513f0.  

This contain an addition of a field, `server_task_type`,  and a virtual function in `server_task_result`. The idea being that the dynamic cast like this can be avoided:
```c++
GGML_ASSERT(dynamic_cast<server_task_result_slot_save_load*>(result.get()) != nullptr);
```
In favor of checks like this:
```c++
        GGML_ASSERT(result.get() != nullptr);                                       
        GGML_ASSERT(result.get()->get_server_task_type() == type);
```
If this looks like alright then I can update the other places in `server.cpp` which also have similar dynamic_casts.

ngxson commented on examples/server/server.cpp:
```diff
@@ -3691,37 +3692,18 @@ int main(int argc, char ** argv) {
+        handle_slot_type(req, res, id_slot, SERVER_TASK_TYPE_SLOT_RESTORE);
     };
 
     const auto handle_slots_erase = [&ctx_server, &res_error, &res_ok](const httplib::Request & /* req */, httplib::Response & res, int id_slot) {
```
IMO we should also extend the function to be reused by `handle_slots_erase`. The only thing need to adapt is that `SERVER_TASK_TYPE_SLOT_ERASE` ignores `filename`

danbev replied:
I'll take a look at this too.

danbev replied:
I've update this in https://github.com/ggerganov/llama.cpp/pull/11535/commits/2d2d07618e8f31b2cb18a0ea2e68869be6d2163c

danbev committed:
squash! server : add handle_slot_type lambda

Rename handle_slot_type to handle_slot_impl.

danbev committed:
server : extract handle_slot_erase to handle_slot_impl

This commit moves the code for handling slot erase to the
handle_slot_impl function. This is done to avoid code duplication and to
make the code more readable.

danbev committed:
server : add server_task_type field to server_task_result

This commit adds a server_task_type field to the server_task_result
struct. This field is used to identify the type of the server task.

The motivation for adding this is that it might allow us to avoid using
dynamic_cast's when checking the type of the server_task_result.
For example, this could then be replaced with checks like this:
```c++
        GGML_ASSERT(result.get() != nullptr);
        GGML_ASSERT(result.get()->get_server_task_type() == type);
```

danbev committed:
squash! server : add server_task_type field to server_task_result

Add missing override keyword.

danbev committed:
squash! server : add server_task_type field to server_task_result

Remove break statement from switch case SERVER_TASK_TYPE_NONE.

slaren commented:
I have cancelled the server workflow since it has been running for 2 hours and seems to be stuck.

danbev commented:
> I have cancelled the server workflow since it has been running for 2 hours and seems to be stuck.

Sorry about that, I'll look into this.

danbev committed:
squash! server : add server_task_type field to server_task_result

Add a check for the server_task_type in the server.cpp file. This is to
ensure that the server_task_type is correct when the server_task_type is
SERVER_TASK_TYPE_SLOT_RESTORE. This is because the server_task_type is
SERVER_TASK_TYPE_SLOT_SAVE when the server_task_type is
SERVER_TASK_TYPE_SLOT_RESTORE.

danbev commented:
Closing as this is change is not worth the effort involved (mine or any reviewers).
